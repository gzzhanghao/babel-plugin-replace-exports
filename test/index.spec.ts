import * as path from 'upath';
import * as babel from '@babel/core';
import plugin, { PluginOptions } from '../src';

interface TransformOptions {
  code: string
  options: PluginOptions
  filename?: string
}

function transform(options: TransformOptions) {
  return babel.transformSync(options.code, {
    configFile: false,
    filename: options.filename || path.resolve('test.js'),
    plugins: [[plugin, options.options]],
  });
}

describe('plugin', () => {
  it('transforms', () => {
    const res = transform({
      code: 'export const value = 1',
      options: { factory: 'foo' },
    });
    expect(res.code).toMatchSnapshot();
  });

  it('includes', () => {
    const res = transform({
      code: 'export const value = 1',
      options: {
        factory: 'foo',
        includes: ['foo.js'],
      },
    });
    expect(res.code).toMatchSnapshot();
  });

  it('transform class declarations', () => {
    const res = transform({
      code: 'export class A {}',
      options: { factory: 'foo' },
    });
    expect(res.code).toMatchSnapshot();
  });

  it('transform function declarations', () => {
    const res = transform({
      code: 'export function foo() {}',
      options: { factory: 'foo' },
    });
    expect(res.code).toMatchSnapshot();
  });

  it('transform exports', () => {
    const res = transform({
      code: `
        const foo = 1
        export { foo }
      `,
      options: { factory: 'foo' },
    });
    expect(res.code).toMatchSnapshot();
  });

  it('transform with factory name', () => {
    const res = transform({
      code: 'export const foo = 1',
      options: {
        factory: 'foo',
        factoryImportName: 'create',
      },
    });
    expect(res.code).toMatchSnapshot();
  });

  it('map filename', () => {
    const res = transform({
      code: 'export const foo = 1',
      options: {
        factory: 'foo',
        mapFilename: (v) => v.replace(/\.(js|ts)x?$/, ''),
      },
    });
    expect(res.code).toMatchSnapshot();
  });

  it('transform specifiers', () => {
    const res = transform({
      code: `
        const foo = 'foo';
        export { foo }
      `,
      options: {
        factory: 'foo',
      },
    });
    expect(res.code).toMatchSnapshot();
  });

  it('transform default export', () => {
    const res = transform({
      code: 'export default foo',
      options: {
        factory: 'foo',
      },
    });
    expect(res.code).toMatchSnapshot();
  });

  it('reexport', () => {
    const res = transform({
      code: `
        import * as foo from 'foo';
        export { foo as bar }

        export { remote as local } from 'bar'

        export * from 'baz'
      `,
      options: {
        factory: 'foo',
      },
    });
    expect(res.code).toMatchSnapshot();
  });
});
