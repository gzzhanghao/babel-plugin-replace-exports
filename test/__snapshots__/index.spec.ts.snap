// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin includes 1`] = `"export const value = 1;"`;

exports[`plugin map filename 1`] = `
"import { default as _factory } from \\"foo\\";
var _filename = \\"test\\";
export var foo = _factory(_filename, \\"foo\\");"
`;

exports[`plugin reexport 1`] = `
"import { default as _factory } from \\"foo\\";
var _filename = \\"test.js\\";
export * as bar from \\"foo\\";
export { remote as local } from \\"bar\\";
export * from \\"baz\\";"
`;

exports[`plugin transform class declarations 1`] = `
"import { default as _factory } from \\"foo\\";
var _filename = \\"test.js\\";
export var A = _factory(_filename, \\"A\\");"
`;

exports[`plugin transform default export 1`] = `
"import { default as _factory } from \\"foo\\";
var _filename = \\"test.js\\";
export default _factory(_filename, \\"default\\");"
`;

exports[`plugin transform exports 1`] = `
"import { default as _factory } from \\"foo\\";
var _filename = \\"test.js\\";
export var foo = _factory(_filename, \\"foo\\");"
`;

exports[`plugin transform function declarations 1`] = `
"import { default as _factory } from \\"foo\\";
var _filename = \\"test.js\\";
export var foo = _factory(_filename, \\"foo\\");"
`;

exports[`plugin transform specifiers 1`] = `
"import { default as _factory } from \\"foo\\";
var _filename = \\"test.js\\";
export var foo = _factory(_filename, \\"foo\\");"
`;

exports[`plugin transform with factory name 1`] = `
"import { create as _factory } from \\"foo\\";
var _filename = \\"test.js\\";
export var foo = _factory(_filename, \\"foo\\");"
`;

exports[`plugin transforms 1`] = `
"import { default as _factory } from \\"foo\\";
var _filename = \\"test.js\\";
export var value = _factory(_filename, \\"value\\");"
`;
